Minio/ApiEndpoints/BucketOperations.cs

```diff
@@ -201,62 +204,64 @@ public async Task SetVersioningAsync(SetVersioningArgs args, CancellationToken c
     ///     For example, if you call ListObjectsAsync on a bucket with versioning
     ///     enabled or object lock enabled
     /// </exception>
-    public IObservable<Item> ListObjectsAsync(ListObjectsArgs args, CancellationToken cancellationToken = default)
-    {
-        args?.Validate();
-        return Observable.Create<Item>(
-            async (obs, ct) =>
-            {
-                var isRunning = true;
-                var delimiter = args.Recursive ? string.Empty : "/";
-                var marker = string.Empty;
-                uint count = 0;
-                var versionIdMarker = string.Empty;
-                var nextContinuationToken = string.Empty;
-                using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ct);
-                while (isRunning)
+    public async IAsyncEnumerable<Item> ListObjectsAsync(ListObjectsArgs args, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        if (args == null) throw new ArgumentNullException(nameof(args));
+        
+        args.Validate();
+
+        var goArgs = new GetObjectListArgs()
+            .WithBucket(args.BucketName)
+            .WithPrefix(args.Prefix)
+            .WithDelimiter(args.Recursive ? string.Empty : "/")
+            .WithVersions(args.Versions)
+            .WithMarker(string.Empty)
+            .WithListObjectsV1(!args.UseV2)
+            .WithHeaders(args.Headers)
+            .WithVersionIdMarker(string.Empty);
+
+        XNamespace ns = "http://s3.amazonaws.com/doc/2006-03-01/";
+        var tag = ns + (args.Versions ? "Version" : "Contents");
+
+        while (true)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            
+            var requestMessageBuilder = await this.CreateRequest(goArgs).ConfigureAwait(false);
+            using var responseResult = await this.ExecuteTaskAsync(ResponseErrorHandlers, requestMessageBuilder, cancellationToken: cancellationToken) .ConfigureAwait(false);
+
+            if (responseResult.StatusCode != HttpStatusCode.OK)
+                throw new ErrorResponseException($"HTTP status-code {responseResult.StatusCode:D}: {responseResult.StatusCode}", responseResult);
+
+#if NET2_0_OR_GREATER
+            var root = await XDocument.LoadAsync(responseResult.ContentStream, LoadOptions.None, cancellationToken).ConfigureAwait(false);
```

Comment:
Make use of async XML loading in .NET core (not available in legacy .NET Framework). Also note that XML reading is done directly on the stream and not using an intermediate string.